packages/astro/src/@types/astro.ts

```diff
@@ -1555,10 +1555,22 @@ export interface AstroUserConfig {
  */
 export type InjectedScriptStage = 'before-hydration' | 'head-inline' | 'page' | 'page-ssr';
 
+/**
+ * IDs for different priorities of injected routes and redirects:
+ * - "override": Override any file-based project route in case of conflict. Conflicts within override
+ *   routes are resolved by the same rules as file-based routes.
+ * - "merge": Merge with discovered file-based project routes, behaving the same as if the route 
+ *   was defined as a file in the project.
+ * - "defer": Defer to any file-based project route in case of conflict. Conflicts within defer
+ *   routes are resolved by the same rules as file-based routes.
+ */
+export type RoutePriorityOverride = 'override' | 'normal' | 'defer';
```

Comment:
Is the plan for this to be user-configurable @Fryuni ? If so, where/how? 

If in `astro.config.mjs` then this file will also need a corresponding `@docs` entry.

Also reminder that you will also need an accompanying PR to update: https://docs.astro.build/en/core-concepts/routing/#route-priority-order 

(FYI It's OK if this update doesn't entirely duplicate the config option, but instead just links to it, if you're also adding it to config reference. It might be most helpful if this sections lists out the default, then links to the configurable option for "fine-tuning")

Reply:
On redirects this is user-configurable, on routes this is configurable for integrations.

Reply:
I don't think injected routes should have the ability to decide how they should be prioritized. I think the only options should be "normal" (ie, the same as if they were file-based routes) and maybe something called "legacy" (the old behavior, only for backwards compatibility, to be removed in next major).

Reply:
I think the customization is positive for integrations.

For example, Starlight injects a 404 route. With this change it could be injected with `priority: 'defer'` allowing projects using Starlight to define a custom 404 page. As it is now this is impossible, Starlight will always override the 404 page.

Also, if we change to have a "normal" and "legacy", I'd expect the "normal" to be the default. But that would be a breaking change. I'm not sure how much of an impact this would be on integrations in the wild, but IMHO it would require a major version by itself.

Reply:
> I think the customization is positive for integrations.

That's a fact, and I agree with your point. However, we risk not providing predictability to the end users, who need to learn what integrations do. While it's OK to change. In the prioritization logic, we also make sure that the user is protected from what integrations do. 

I believe we can progressively provide integrations more control, although we should first fix the predictability of the prioritization.

Reply:
@Fryuni It should be possible for Starlight to define 404 via middleware. I think a lot of what Starlight does might be better via middleware. Currently middleware is a little limited and can't do something like 404, we are in the process of fixing that though.

Reply:
Rethinking about this, if an integration wants to provide a route that can be overridden by the project it is on, it could just allow disabling the route. If Starlight wants to allow overriding the 404 page it could have a `notFoundRoute: false` and not even call `injectRoute` in that case.

Considering that I don't see much advantage of allowing these different priorities.

Reply:
I was hoping to avoid exactly that kind of option with the priority system.

UX with a `notFoundRoute: false` option:

1. User creates a `404.astro` 
2. They get an error about route collisions (see https://github.com/withastro/starlight/issues/1080)
3. They need to work out Starlight’s option exists — likely not something I can add helpful logging for in Starlight because the error is Astro’s
4. Maybe they set `notFoundRoute: false` if they find it, maybe they assume this isn’t supported.

UX with a lower route priority when injecting Starlight’s 404:

1. User creates a `404.astro`. It works!